### DISCUSSION TITLE
jslt for loop capability

### DISCUSSION THREAD

**Discussion by brh** [schibsted/jslt#191] [Q&A] [answered]
How can I use the for loop to concat a string as opposed to making an array.
JSON:
`{
  "x": [
    {
      "y": "y1"
    },
    {
      "y": "y2"
    },
    {
      "y": "y3"
    }
  ]
}`

Should turn into this
{
"arbitraryField":"y1,y2,y3"
}

**Comment by brh** [comment]
Oops, I found join, although I still am having trouble getting it to work.

**Reply by larsga** [reply]
What's the problem?

**Reply by brh** [reply]
Ok closer to the real world example.
`{
  "cartId": "6795708823149918171",
  "cartItems": [
    {
      "fulfillments": [
        {
          "fulfillmentType": "PARCEL",
        },
        {
          "fulfillmentType": "PICKUP",
        },
        {
          "fulfillmentType": "TRUCK",
        }
      ],

  "createdOn": "2021-05-05T15:09:14Z"
}`

Then I wrote a template like this:
`{
  "&&events": [for (.cartItems) string(.fulfillments.fulfillmentType)]
}`

The response I get is 
`{"&&events":["null"]}`

So once I get the correct values of ["PARCEL","PICKUP","TRUCK"]
I want it to be 
{"&&events":"xxx=PARCEL;xxx=PICKUP,xxx=TRUCK"}

**Reply by larsga** [reply]
Is there really an array inside `cartItems`? If so, you're not accounting for it.

If I were you I'd start out with real data and the playground, and write an expression there that gives `["PARCEL","PICKUP","TRUCK"]` as the output, and only then try to put together the rest.

**Reply by brh** [reply]
Ok, but yes there can be multiple cartItems

**Comment by larsga** [comment, marked as answer]
Given this input:

```
{"cartId": "6795708823149918171",
"cartItems": [ {
  "fulfillments": [ {"fulfillmentType": "PARCEL"},
                    {"fulfillmentType": "PICKUP"},
                    {"fulfillmentType": "TRUCK"}]},
  {
  "fulfillments": [ {"fulfillmentType": "OTHER"}]}
],
"createdOn": "2021-05-05T15:09:14Z"
}
```

you can get the output you want with:

```
let tmp = flatten([for(.cartItems) .fulfillments])

{"&&events" : join([for ($tmp) "xxx=" + .fulfillmentType], ";")}
```

Having to use flatten and multiple for loops is not really ideal, so I think we do need to add an array traversal operator (already planned in issue #183).

**Reply by brh** [reply]
That did it thanks!!!
